### PR TITLE
image: mkosi from nixpkgs

### PR DESCRIPTION
**Pull request by malt3**
### Context

This is a large step towards reproducible OS image builds. With this change, mkosi and all required tools are installed with nix.
In a follow up step, we will pin all RPMs required to build a Fedora image.

### Proposed change(s)

- use rules_nixpkgs
- provide toolchains from nixpkgs (allows running `bazel build` directly on NixOS)
- mkosi ruleset
- refactor image build for new mkosi version
- reimplement measurement precalculation

### How to review

Since this PR is quite large, it is probably a good idea to not have everyone review it in full.
Instead, review a part that you are the expert of and comment what you reviewed.
Also, please test if you can build an image on your (Linux) workstation.

### How to test

- When creating a cluster, you can use the following debug image created by the pipeline: `ref/feat-image-nix-mkosi-toolchain/stream/debug/v2.12.0-pre.0.20230926093438-08d3e02d2353`
- On NixOS, add `common --config=nix` to `.bazeloverwriterc` and build images (see image/README.md)
- On non-NixOS
  - notice how image build fails without nix installed
  - notice how the rest still works as expected (go binaries)
  - install nix using the [determinate systems installer](https://github.com/DeterminateSystems/nix-installer)
  - add `common --config=nix` to `.bazeloverwriterc`
  - run `bazel clean --expunge` and build images (see image/README.md)

### Checklist

- [x] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [x] Is PR title adequate for changelog?
- [x] Link to Milestone
- [ ] e2e tests
  - [x] [GCP](https://github.com/edgelesssys/constellation/actions/runs/6314268521)
  - [x] [Azure](https://github.com/edgelesssys/constellation/actions/runs/6312353216)
  - [ ] [AWS](https://github.com/edgelesssys/constellation/actions/runs/6312337834) 
- [x] [build pipeline in CI](https://github.com/edgelesssys/constellation/actions/runs/6311065034)

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        ];
+      }));
+
+    in
+    {
+      packages.mkosi = mkosiDev;
+
+      devShells.default = import ./nix/shells/default.nix { pkgs = pkgsUnstable; };
+
+      formatter = nixpkgsUnstable.legacyPackages.${system}.nixpkgs-fmt;
```
Comment (katexochen) [inline flake.nix:59]: We should add a formatter/linter/instantiate workflow to the CI (not in this PR, but soon).

File: dev-docs/workflows/build-develop-deploy.md [hunk 2]
```diff
@@ -6,13 +6,22 @@ Prerequisites:
...
 
 ## Prequisites
 
 ### Linux
 
+* Packages on NixOS or with Nix installed (use flake.nix in this repo):
+
+  ```sh
+  # add "--config=nix" to your .bazeloverwriterc if you want to get nix compatible toolchains
+  echo "build --config=nix" >> .bazeloverwriterc
```
Comment (msanft) [inline dev-docs/workflows/build-develop-deploy.md:21]: Should we add `build --config=nix` or `--config=nix`?

**Review by msanft — COMMENTED** [review]
Image builds complete successfully on NixOS after a cache clean, nice work!🤟

File: image/initrd/mkosi.skeleton/usr/sbin/aws-nvme-disk [hunk 3]
```diff
@@ -9,6 +9,14 @@ shopt -s extglob nullglob inherit_errexit
 AWS_STATE_DISK_DEVICENAME="sdb"
 AWS_STATE_DISK_SYMLINK="/dev/${AWS_STATE_DISK_DEVICENAME}"
 
+function onError {
+  echo "Failed to symlink state disk"
+  sleep 2 # give the serial console time to print the error message
+  exit $? # exit with the same error code
```
[the remark this answers is not in the document]
Comment (malt3) [reply image/initrd/mkosi.skeleton/usr/sbin/aws-nvme-disk:15]: Correct. Interestingly, this code was copied over from the `prepare-state-disk.sh`, where it was even more broken (did not use trap). I'll just return a hard coded exit code instead.

**Review by daniel-weisse — APPROVED** [review]
Image build worked for me on Fedora 37

**Comment by malt3** [comment]
> Should we document the change to disabled secure boot on GCP?

I'll document this and also update the docs with the boot measurement table

**Pull request merged**